File: application/src/com/yuyan/wx/login/data/Result.java
package com.yuyan.wx.login.data;

public class Result<T> {
    private Result() {}

    public final static class Success<T> extends Result<T> {
        private final T data;

        public Success(T data) {
            this.data = data;
        }

        public T getData() {
            return data;
        }
    }

    @SuppressWarnings("rawtypes")
    public static final class Error extends Result {
        private final Exception exception;

        public Error(Exception exception) {
            this.exception = exception;
        }

        public Exception getException() {
            return exception;
        }
    }
}

File: app/harp/src/com/yuyan/harp/data/model/Result.kt
package com.yuyan.harp.data.model


open class Result<T> // hide the private constructor to limit subclass types (Success, Error)
private constructor() {
    override fun toString(): String {
        if (this is Success<*>) {
            val success =
                    this as Success<*>
            return "Success[data=" + success.data.toString() + "]"
        } else if (this is Error) {
            val error = this as Error
            return "Error[exception=" + error.error.toString() + "]"
        }
        return ""
    }

    // Success sub-class
    class Success<T>(val data: T) : Result<T>()

    // Error sub-class
    class Error(val error: Exception) : Result<Any?>()
}